logstash-core/src/main/java/org/logstash/common/BufferedTokenizer.java

```diff
@@ -0,0 +1,181 @@
+    private final Iterable<String> iterable;
+    private Integer sizeLimit;
+
+    static abstract class IteratorDecorator<T> implements Iterator<T> {
+        protected final Iterator<String> iterator;
+
+        IteratorDecorator(Iterator<String> iterator) {
+            this.iterator = iterator;
+        }
+
+        @Override
+        public boolean hasNext() {
+            return iterator.hasNext();
+        }
+    }
+
+    static class ValueLimitIteratorDecorator extends IteratorDecorator<String> {
+        private final int limit;
+
+        ValueLimitIteratorDecorator(Iterator<String> iterator, int sizeLimit) {
+            super(iterator);
+            this.limit = sizeLimit;
+        }
+
+        @Override
+        public String next() {
+            String value = iterator.next();
+            if (value.length() > limit) {
+                throw new IllegalStateException("input buffer full, consumed token which exceeded the sizeLimit " + limit);
+            }
+            return value;
+        }
+    }
+
+    static class DataSplitter implements Iterator<String> {
+        private final String separator;
+        private int currentIdx = 0;
+        private int nextSeparatorIdx = -1;
+        private final StringBuilder accumulator = new StringBuilder();
+
+        DataSplitter(String separator) {
+            this.separator = separator;
+        }
+
+        @Override
+        public synchronized boolean hasNext() {
+            return matchNextSeparatorIdx();
+        }
+
+        @Override
+        public synchronized String next() {
+            if (!matchNextSeparatorIdx()) {
+                throw new NoSuchElementException();
+            }
+
+            String token = accumulator.substring(currentIdx, nextSeparatorIdx);
+            currentIdx = nextSeparatorIdx + separator.length();
+            nextSeparatorIdx = -1;
+            return token;
+        }
+
+        /**
+         * Used to retrieve the index of the next token just one time. It saves into the nextSeparatorIdx
+         * and return true if a next token is present.
+         * Updates internal state for tracking.
+         *
+         * @return true iff a next complete token is available.
+         */
+        private boolean matchNextSeparatorIdx() {
+            if (nextSeparatorIdx == -1) {
+                nextSeparatorIdx = accumulator.indexOf(separator, currentIdx);
+            }
+            // clean up accumulator if no next separator found
+            if (nextSeparatorIdx == -1 && currentIdx > 0) {
+                cleanupAccumulator();
+            }
+            return nextSeparatorIdx != -1;
+        }
+
+        private void cleanupAccumulator() {
+            accumulator.delete(0, currentIdx);
+            currentIdx = 0;
+        }
+
+        public void append(String data) {
+            accumulator.append(data);
+        }
+
+        public String flush() {
+            final String flushed = accumulator.substring(currentIdx);
```

Comment:
I'm surprised that we can accumulate many times the size limit, and a flush yields us a string that is many times the size limit.

~~~ java

    @Test
    public void givenBuffersWithoutTerminatorsBeyondSizeLimitFlush() {
        final BufferedTokenizer sut2 = new BufferedTokenizer("\n", 1000);
        for (int i = 0; i < 100000; i++) {
            Iterable<String> extract = sut2.extract("this is a long string that does not include the separator!");
            assertFalse(extract.iterator().hasNext());
        }
        // by this point we have hundreds of times the limit in the accumulator
        assertThrows(IllegalStateException.class, () -> sut2.flush());
    }
~~~